### ISSUE TITLE
Can't Query Parent By Related Child

### ISSUE THREAD

**Issue by noamby** [django-salesforce/django-salesforce#261]
Trying to fix a URI too long error when fetching a parent object by id from related child
For example, if I have:
```
class A():
    email

class B():

class AToB():
    a = models.ForeignKey(A, models.DO_NOTHING, custom=True,
                                related_name="a_to_bs", blank=True, null=True, db_column='A__c')
    b = models.ForeignKey(B, models.DO_NOTHING, custom=True,
                                related_name="a_to_bs", blank=True, null=True, db_column='B__c')
```
And I want to get all B object related to a specific A email

This results in URI too long if I have enough relations:
```
email = 'email@some.com'
a_to_b_list = AToB.objects.filter(a__email=email)
b_id_list = [a_to_b.b.id for a_to_b in a_to_b_list]
b_list = B.objects.filter(id__in=b_id_list)
```

This results in query on the related model (AToB model) and fetching the parent
So if I have 2 relations to one B object I will get a list containing 2 B objects with the same id
Also, if I add `order_by` this would result in a `MALFORMED_QUERY` because the order by field would be B.{field_name} and the query is on the child object
```
email = 'email@some.com'
b_list = B.objects.filter(a_to_bs__a__email=email)
```

The SOQL I would expect is:
```
SELECT B.Id FROM B WHERE B.Id IN (SELECT B__c FROM AToB WHERE AToB.A__r.Email = 'email@some.com')
```
What I get is this SOQL for related object with parent only fields that return a list of parent objects:
```
SELECT AToB.B.Id FROM AToB WHERE AToB.A__r.Email = 'email@some.com'
```

**Comment by hynekcer** [comment]
I agree that this is a problem and that is not nice for you.

1) Solution

    b_id_set = set(AToB.objects.filter(a__email=email).values_list('b_id', flat=True)
    b_list = B.objects.filter(id__in=b_id_set)

2) Comments to your text:

Your expression `[a_to_b.b.id for a_to_b in a_to_b_list]` will be probably slow, while a similar expression `[a_to_b.b_id for a_to_b in a_to_b_list]` is fast.

You have two relations to one B object, but you get **one** B object by the expression `B.objects.filter(id__in=b_id_list)`. The word `set` is not necessary and it works less effectively also with `list`. It could be important `len(b_id_list)` is more than approximately 400, while `len(set(b_id_set))` is less than 350.

The word "Custom" is not important and the problem is very similar with any ManyToMany relationship, e.g. with "Campaign <- CampaignMember -> Contact".

3) Resume for development: 3 issues

* "URI too long" can be fixed a little (by SOQL in a body of the request, not in URL) until it is limited by the maximal number of parameters.

* `order_by()` can be fixed by more complicatedly probably using a compiler similar to a compiler used for a `SELECT` clause in queries with relationships.

* `DISTINCT` clause is not supported by SOQL. The same can be sometimes achieved by "GROUP BY", but some fields are not [**groupable** in SOQL](https://developer.salesforce.com/docs/atlas.en-us.224.0.api.meta/api/sforce_api_calls_describesobjects_describesobjectresult.htm) My opinion is that a satisfying fix is not possible in the framework and a workaround in user code is easy.

**Comment by noamby** [comment]
@hynekcer thanks for the quick update!
I renamed the issue like you asked

I agree the best solution is 1 but unfortunately I still get a URI too long error 😞 

Regarding this solution:
```
email = 'email@some.com'
b_list = B.objects.filter(a_to_bs__a__email=email)
```
Do you know why it would query the child object and not the parent?

**Comment by hynekcer** [comment]
* 1) "URI too long" is now fixed by PR #263

* 3) I tried that if your example is rewritten by "OpportunityContactRole" and used a 'default' database, then it is compiled to `... FROM Opportunity JOIN OpportunityContactRole ... JOIN Contact...` and repeated parent objects are probably possible.

Please verify it also on your model and rewrite the queryset so that it works on a default non-salesforce database. Probably by `Subquery()`paragraph  [Limiting the subquery to a single row](https://docs.djangoproject.com/en/3.0/ref/models/expressions/#s-limiting-a-subquery-to-a-single-column).

Builtin models are preferable for tests because a custom model can not be so easily reproduced. E.g. ManyToMany relationship through a table OpportunityContactRole between tables Opportunity and Contact is also a part of `salesforce.testrunner.example.models`.

**Comment by hynekcer** [comment]
**The solution** is easy with a subquery for the used association table and without any fix in the framework:

``` python
B.objects.filter(pk__in=AtoB.objects.filter(a__email='a@example.com').values('b'))
```
It is compiled and executed as one SOQL request:
``` sql
SELECT B__c.Id,... FROM B__c WHERE B__c.Id IN (
    SELECT AtoB__c.B__c FROM AtoB__c WHERE AtoB__c.A__r.Email__c = 'a@example.com'
)
```
---
Rewritten to use only builtin models:
``` python
associations = OpportunityContactRole.objects.filter(contact__email='a@example.com')
qs = Opportunity.objects.filter(pk__in=associations.values('opportunity'))

print(str(qs.query))
```
Result:
``` sql
SELECT Opportunity.Name,... FROM Opportunity
WHERE Opportunity.Id IN (
    SELECT OpportunityContactRole.OpportunityId FROM OpportunityContactRole
    WHERE OpportunityContactRole.Contact.Email = 'a@example.com'
)
```
Other variants:
``` python
# var 2
qs = Opportunity.objects.filter(pk__in=associations.values('opportunity'))

# var 3
from django.db.models import Subquery
qs = Opportunity.objects.filter(pk__in=Subquery(associations.values('opportunity')))
```
All three variants are compiled to the same SOQL and have the same result. The last most explicit variant can be sometimes useful.

---
**Resume:**
* I'm going to write a separate issue for the bug `order_by(...)` (and similar methods combined with parent to child relationship) to can close this solved issue
* To write a Wiki or a cheat sheet about ManyToMany
* A commit with two tests has been written

**Comment by noamby** [comment]
@hynekcer thanks!

The fix you did for "URI too long" would help a lot
Subscribed to the new issue you opened to get a fix :)

**Comment by hynekcer** [comment]
Everything is fixed now and described more in #264.

The only feature unexpected by your issue report is that the same parent row can be repeated if more ManyToMany rows exist for it. This is similar standard databases where the queryset is compiled by "INNER JOIN" or "LEFT JOIN" and also repeated rows could be in results. However standard SQL databases can use DISTINCT or DISTINCTROW operators, but they have no equivalent in SOQL.
